src/plugins/intel_cpu/src/memory_control.cpp

```diff
@@ -164,49 +301,141 @@ class MemoryManagerStatic : public IMemoryManager {
         }
     }
 
+    static const char* getClassName() {
+        return "MemoryManagerStatic";
+    }
+
 private:
-    MemoryControl::MemoryBlockMap m_blocks;
+    MemoryControl::MemorySolution m_blocks;
     std::vector<MemorySolver::Box> m_boxes;
     std::shared_ptr<MemoryBlockWithRelease> m_workspace;
     size_t m_totalSize = 0;
+    bool reset_flag = true;
 };
 
-class MemoryManageNonOverlapingSets : public IMemoryManager {
+class MemoryManagerNonOverlappingSets : public IMemoryManager {
 public:
-    MemoryManageNonOverlapingSets(std::vector<size_t> syncInds) : m_syncInds(std::move(syncInds)) {}
-    void insert(const MemoryRegion& reg) override {
+#ifdef CPU_DEBUG_CAPS
+    class MemoryBlocks {
+    public:
+        MemoryBlocks() = default;
+        explicit MemoryBlocks(const std::vector<std::vector<ov::MemorySolver::Box>>& groups) {
+            for (auto& group : groups) {
+                m_unique_blocks.push_back(std::make_shared<MemoryBlockWithRelease>());
+                for (auto& box : group) {
+                    m_internalBlocks[box.id] =
+                        std::make_shared<IndividualMemoryBlockWithRelease>(m_unique_blocks.back());
+                }
+            }
+        }
+        auto begin() {
+            return m_internalBlocks.begin();
+        }
+
+        auto end() {
+            return m_internalBlocks.end();
+        }
+        const auto& at(const MemoryControl::MemorySolution::key_type& key) const {
+            return m_internalBlocks.at(key);
+        }
+        const auto& uniqueBlocks() const {
+            return m_unique_blocks;
+        }
+
+    private:
+        std::vector<std::shared_ptr<MemoryBlockWithRelease>> m_unique_blocks;
```

Comment:
It should be possible to count unique_blocks even without this data structure, isn't?
I mean it will be slower for sure, but do we care how fast we collect those statistics?

Reply:
This is also used to calculate the actually allocated memory size. The wrappers stores the memory requested from each tensor, but not the really allocated memory.

Reply:
Just trying to understand whether we can have production code untouched (or maybe refactored?), so we can get all the necessary information without introducing extra logic. But maybe it is not feasible here.

Reply:
Yes, that's true. Technically we can extend the wrapper interface with an accessor to the parent block, then build a set of unique blocks once requested. Theoretically it may help to avoid using two `MemoryBlocks` classes at the cost of more complex data retrieval algo. But anyway, the compile time type should be different as in the case of DEBUG_CAPS build we still need the wrapper.
I don't think we really need to set a goal of leaving the production code untouched, as it doesn't make sense to keep debug information in the production build.

Reply:
From my perspective it would just look cleaner - to only substitute the memory block implementation to achieve memory statistics collection. Like we injected different implementation and achieved what we need. I can imagine using such approach in scope of unit testing.

Reply:
Thank you for highlighting this problem. I made an attempt to simplify the `MemoryManagerNonOverlappingSets` substituting the proxy type in only one place. Hope this helps to make the implementation cleaner.